04.ls/ls.rb

```diff
@@ -90,62 +95,57 @@ def to_file_name(stat, absolute_path)
   stat.symlink? ? "#{file_name} -> #{File.readlink(absolute_path)}" : file_name
 end
 
-def format_file(file_data, maxlength_by_section)
+def format_file(file_data, max_lengths)
   space = ' '
   [
     file_data[:permission], space * 2,
-    file_data[:link_count].rjust(maxlength_by_section[:link]), space,
-    file_data[:owner].ljust(maxlength_by_section[:owner]), space * 2,
-    file_data[:group].ljust(maxlength_by_section[:group]), space * 2,
-    file_data[:byte_size].rjust(maxlength_by_section[:byte_size]),
+    file_data[:link_count].rjust(max_lengths[:link]), space,
+    file_data[:owner].ljust(max_lengths[:owner]), space * 2,
+    file_data[:group].ljust(max_lengths[:group]), space * 2,
+    file_data[:byte_size].rjust(max_lengths[:byte_size]),
     file_data[:time_stamp],
     file_data[:file_name]
   ].join
 end
 
-def fetch_maxlength(stat_files)
-  length_by_section = { link: 0, owner: 0, group: 0, byte_size: 0 }
+def result_max_lengths(stat_files)
+  lengths = { link: 0, owner: 0, group: 0, byte_size: 0 }
 
   stat_files.each do |stat|
-    length_by_section[:link] = stat.nlink.to_s.length if length_by_section[:link] < stat.nlink.to_s.length
-    length_by_section[:owner] = Etc.getpwuid(stat.uid).name.length if length_by_section[:owner] < Etc.getpwuid(stat.uid).name.length
-    length_by_section[:group] = Etc.getgrgid(stat.gid).name.length if length_by_section[:group] < Etc.getgrgid(stat.gid).name.length
-    length_by_section[:byte_size] = stat.size.to_s.length if length_by_section[:byte_size] < stat.size.to_s.length
+    lengths[:link] = stat.nlink.to_s.length if lengths[:link] < stat.nlink.to_s.length
+    lengths[:owner] = Etc.getpwuid(stat.uid).name.length if lengths[:owner] < Etc.getpwuid(stat.uid).name.length
+    lengths[:group] = Etc.getgrgid(stat.gid).name.length if lengths[:group] < Etc.getgrgid(stat.gid).name.length
+    lengths[:byte_size] = stat.size.to_s.length if lengths[:byte_size] < stat.size.to_s.length
   end
-  length_by_section
+  lengths
 end
 
 def to_permission(stat)
-  FILE_TYPE[stat.ftype] + divide_character(stat).map do |section, char|
-    to_uid_or_gid(stat, section, char) || to_sticky(stat, section, char) || PERMISSION[char]
-  end.join
+  to_characters = divide_characters(stat).map do |section, char|
+    to_special_permission(stat, section, char) || PERMISSIONS[char]
+  end
+  FILE_TYPES[stat.ftype] + to_characters.join
 end
 
-def divide_character(stat)
+def divide_characters(stat)
   # mode.to_s(8)の戻り値の文字数は可変なため、末尾から指定しています。
-  stat.mode.to_s(8).slice(-3..-1).chars.map.with_index do |char, idx|
+  stat.mode.to_s(8).slice(-3..-1).chars.each_with_index.to_h do |char, idx|
     case idx
     when 0 then [:user, char]
     when 1 then [:group, char]
     when 2 then [:other, char]
     end
```

Comment:
```suggestion
  hoge = stat.mode.to_s(8)
  {
    user: hoge[-3],
    group: hoge[-2],
    other: hoge[-1]
  }
```

そもそもこんな感じにした方がシンプルだったりしません？

Reply:
`chars`は使うものと思い込んでしまっていました。。